plugins/thunderbolt/fu-thunderbolt-firmware.c

```diff
@@ -0,0 +1,517 @@
+		{ 0x15D3, 3, 2 }, /* AR-C 4C */
+		{ 0x15DA, 3, 1 }, /* AR-C 2C */
+		{ 0x15E7, 3, 1 }, /* TR 2C */
+		{ 0x15EA, 3, 2 }, /* TR 4C */
+		{ 0x15EF, 3, 2 }, /* TR 4C device */
+		{ 0 }
+	};
+	g_autoptr(FuFirmwareImage) img = fu_firmware_image_new (fw);
+
+	/* add this straight away so we can read it without a self */
+	fu_firmware_add_image (firmware, img);
+
+	/* is native */
+	if (!fu_thunderbolt_firmware_read_uint8 (self,
+						 _SECTION_DIGITAL,
+						 FU_TBT_OFFSET_NATIVE,
+						 &tmp, error)) {
+		g_prefix_error (error, "failed to read native: ");
+		return FALSE;
+	}
+	self->is_native = tmp & 0x20;
+
+	self->sections[_SECTION_DIGITAL] = fu_thunderbolt_firmware_read_farb_pointer (self, error);
+	if (self->sections[_SECTION_DIGITAL] == 0)
+		return FALSE;
+
+	/* host or device */
+	if (!fu_thunderbolt_firmware_read_uint8 (self,
+						 _SECTION_DIGITAL,
+						 0x10, &tmp, error)) {
+		g_prefix_error (error, "failed to read is-host: ");
+		return FALSE;
+	}
+	self->is_host = tmp & (1 << 1);
+
+	/* device ID */
+	if (!fu_thunderbolt_firmware_read_uint16 (self,
+						  _SECTION_DIGITAL,
+						  0x5,
+						  &self->device_id,
+						  error)) {
+		g_prefix_error (error, "failed to read device-id: ");
+		return FALSE;
+	}
+
+	/* this is best-effort */
+	for (guint i = 0; hw_info_arr[i].id != 0; i++) {
+		if (hw_info_arr[i].id == self->device_id) {
+			self->gen = hw_info_arr[i].gen;
+			self->ports = hw_info_arr[i].ports;
+			break;
+		}
+	}
+	if (self->ports == 0 && self->is_host) {
+		g_set_error (error,
+			     FWUPD_ERROR,
+			     FWUPD_ERROR_NOT_SUPPORTED,
+			     "Unknown controller");
+		return FALSE;
+	}
+
+	/* read sections from file */
+	if (!fu_thunderbolt_firmware_read_sections (self, error))
+		return FALSE;
+	if (fu_thunderbolt_firmware_missing_needed_drom (self)) {
+		g_set_error_literal (error,
+				     FWUPD_ERROR,
+				     FWUPD_ERROR_READ,
+				     "Can't find required FW sections");
+		return FALSE;
+	}
+
+	/* vendor:model */
+	if (self->sections[_SECTION_DROM] != 0) {
+		if (!fu_thunderbolt_firmware_read_uint16 (self,
+							  _SECTION_DROM,
+							  0x10,
+							  &self->vendor_id,
+							  error)) {
+			g_prefix_error (error, "failed to read vendor-id: ");
+			return FALSE;
+		}
+		if (!fu_thunderbolt_firmware_read_uint16 (self,
+							  _SECTION_DROM,
+							  0x12,
+							  &self->model_id,
+							  error)) {
+			g_prefix_error (error, "failed to read model-id: ");
+			return FALSE;
+		}
+	}
+
+	/* has PD */
+	if (self->sections[_SECTION_ARC_PARAMS] != 0) {
+		guint32 pd_pointer = 0x0;
+		if (!fu_thunderbolt_firmware_read_uint32 (self,
+							  _SECTION_ARC_PARAMS,
+							  0x10C,
+							  &pd_pointer,
+							  error)) {
+			g_prefix_error (error, "failed to read pd-pointer: ");
+			return FALSE;
+		}
+		self->has_pd = fu_thunderbolt_firmware_valid_pd_pointer (pd_pointer);
+	}
+
+	if (!self->is_host) {
+		/* TODO: is this to be multiplied by a multiplier, e.g. 0x10000? */
```

Comment:
@YehezkelShB do you know?

Reply:
There is a reason why it was just one another location in the comparisons done for the images. This is just an "enum"-like value. There is no point with trying to treat it as a "real" size.

Reply:
I've adjusted the comment to make it clear that we shouldn't actually use the value here but just match it between images.